### DISCUSSION TITLE
插件模块的开发

### DISCUSSION THREAD

**Discussion by lengkristy** [netease-im/NIM_Duilib_Framework#492] [General]
请问一下这个有没有模块插件的开发教程，因为UI里面有很多的模块，想做成动态loadlibrary的方式加载模块dll，不用全部将UI写到一个exe里面，这样也方便功能模块的控制

**Comment by nmgwddj** [comment]
示例中没有做过这样的尝试。但从开发角度考虑这应该不难实现。可以尝试将 ui_component 中部分组件编译为动态库，在需要时动态链接。

**Reply by lengkristy** [reply]
我自己做了一个dll，但是通过主程序加载 在调用创建函数Create的时候 就报错了，空指针错误，我用ui::GlobalManager初始化了一下，然后又报错：
Run-Time Check Failure #0 - The value of ESP was not properly saved across a function call.  This is usually a result of calling a function declared with one calling convention with a function pointer declared with a different calling convention.

我用ui_component编译成动态库试试

**Reply by nmgwddj** [reply]
动态库区别于静态链接，一些全局资源跨模块之间一旦被隐藏时无法加载的，设计上应尽量考虑避免在动态库中依赖全局资源，如必不得已必须使用，需要通过接口的方式将全局资源注入到动态库空间。
或者您可以将 duilib 作为基础动态库，让 app 与插件同时共享一份资源，这样更为合理。

**Reply by lengkristy** [reply]
好的，明白了，谢谢

**Comment by lengkristy** [comment]
在生成动态库的过程一堆链接错误，解决了链接错误之后，duilib项目编译没问题了，但是没有预期生成对应的dll文件，base项目都生成了dll文件，感觉动态库的bug有点多。
生成日志：
1>CircleProgress.cpp
1>VBox.cpp
1>TileBox.cpp
1>TabBox.cpp
1>正在生成代码...
1>正在编译...
1>HBox.cpp
1>ChildBox.cpp
1>AnimationPlayer.cpp
1>AnimationManager.cpp
1>正在生成代码...
1>C:\Program Files (x86)\MSBuild\Microsoft.Cpp\v4.0\V120\Microsoft.CppBuild.targets(1193,5): warning MSB8012: TargetPath(E:\source_code\vs_work\cpp\NIM_Duilib_Framework-master\duilib\..\libs\duilib_d.dll) does not match the Linker's OutputFile property value (E:\source_code\vs_work\bin\duilib_d.dll). This may cause your project to build incorrectly. To correct this, please make sure that $(OutDir), $(TargetName) and $(TargetExt) property values match the value specified in %(Link.OutputFile).
1>C:\Program Files (x86)\MSBuild\Microsoft.Cpp\v4.0\V120\Microsoft.CppBuild.targets(1197,5): warning MSB8030: The linker switch "Minimum Required Version" requires "SubSystem" to be set.  Without "SubSystem", the "Minimum Required Version" would not be passed to linker and could prevent to the output binary from running on older Operating Systems.
1>AnimationManager.obj : warning LNK4075: 忽略“/EDITANDCONTINUE”(由于“/OPT:LBR”规范)
1>duilib.vcxproj -> E:\source_code\vs_work\cpp\NIM_Duilib_Framework-master\duilib\..\libs\duilib_d.dll
1>已完成生成项目“duilib.vcxproj”的操作。
========== “生成”: 1 成功，0 失败，0 更新，0 已跳过 ==========
![image](https://github.com/netease-im/NIM_Duilib_Framework/assets/37865070/0eb29e96-1e18-4688-afcc-e9e022bf1a3b)

去目录下面看，没有对应的dll.
![image](https://github.com/netease-im/NIM_Duilib_Framework/assets/37865070/1ea0dbd5-411f-4df3-8bb0-39d41bd14a37)

找遍了全网，好像都没看到这个库有生成dll的文章，只有原生duilib才有生成dll，而且我发现base项目里面的宏定义BASE_EXPORT，如果让他生效，也会有一堆链接错误，如果不生效，duilib项目依赖的这个base项目里面类没有导出，也造成了链接错误，感觉dll比较混乱。

**Comment by lengkristy** [comment]
搞定了，这个配置的生成路径有问题，建议官方把生成dll的vc配置弄一下，不然新人入手太多坑了
